src/hip/test/radixSort_t.cu

```diff
@@ -140,8 +140,9 @@ void go(bool useShared) {
         auto sh = sizeof(uint64_t) - NS;
         sh *= 8;
         auto shorten = [sh](T& t) {
-          auto k = (uint64_t*)(&t);
```

Comment:
is this approach used only in the test ? or also in some of the application code ?

Reply:
The `shorten` function is only used in the radixSort_t test in the cuda* ports and the hip port.  The radixSort function is used in plugin-PixelVertexFinding/gpuSortByPt2.h, and it doesn't seem that any such bit-level processing of the numbers is needed there.

Reply:
(apologies for long delay)

I'm trying to understand what this lambda is supposed to do
```cpp
        auto sh = sizeof(uint64_t) - NS;
        sh *= 8;
        auto shorten = [sh](T& t) {
          auto k = (uint64_t*)(&t);
          *k = (*k >> sh) << sh;
        };
```
It gets run for `T ` with `sizeof` equal to that of `int8_t, int16_t, int32_t, int64_t`, leading to `sh` values of 56, 48, 32. The code appears to zero the lowest `sh` number of bits, i.e. the portion of `T` of the `uint64_t` piece of memory. Adding a printout here (for `cuda` program) confirms that whenever `sh != 0`, both `k1` and `k2` are `0`.

@VinInn, could you help here with the intention of this lambda? (the code is exactly the same in CMSSW)

Reply:
to test sorting? (one way to speed up radix sort is to consider only the MSBs.  for instance in ptsort only the 16MSBs are used. 
most probably this lambda is used to verify that the result is the same (I need to check)

Reply:
> most probably this lambda is used to verify that the result is the same (I need to check)

Right, but setting the compared-to values to zero for all types `T` shorter than `uint64_t` doesn't sound very useful for that.

Reply:
ok. most probably was written for 64bit values and never fix it...

Reply:
Here is a version that keep GCC 11 happy and seems to work as intended for arbitrary types:
```c++
auto shorten = [](T& t) {
    // byte representation of t
    char* bytes = reinterpret_cast<char*>(&t);
    // bytes to zero out
    const int zeroes = static_cast<int>(sizeof(T)) - NS;
    // zero out the least significant bytes (assuming a little endian architecture)
    static_assert(__BYTE_ORDER__ == __ORDER_LITTLE_ENDIAN__, "This test assumes a little-endian architecture");
    for (int i = 0; i < zeroes; ++i) {
        bytes[i] = 0x00;
    }
};
```

Reply:
And here is a version that is more similar to the original, and avoids the loop, but adds a lot of boilerplate to keep the compiler happy about "aliasing rules":
```c++
template <int N>
struct uintN;

template <>
struct uintN<8> {
    using type = uint8_t;
};

template <>
struct uintN<16> {
    using type = uint16_t;
};

template <>
struct uintN<32> {
    using type = uint32_t;
};

template <>
struct uintN<64> {
    using type = uint64_t;
};

template <int N>
using uintN_t = typename uintN<N>::type;

template <typename T>
using uintT_t = uintN_t<sizeof(T) * 8>;


template <typename T>
union safe_uint_cast {
    T t;
    uintT_t<T> u;
};

template <int N, typename T, typename SFINAE = std::enable_if_t<N <= sizeof(T)>>
void shorten(T& t) {
    const int shift = 8 * (sizeof(T) - N);
    safe_uint_cast<T> c;
    c.t = t;
    c.u = c.u >> shift << shift;
    t = c.t;
}
```

Reply:
PR'ed in #209.